Src/WindowsServer/WindowsServer.Shared/ConfigurationTelemetryInitializer.cs

```diff
@@ -0,0 +1,114 @@
+        /// </summary>
+        public ConfigurationTelemetryInitializer()
+        {
+            WindowsServerEventSource.Log.TelemetryInitializerLoaded(this.GetType().FullName);
+            AppServiceEnvironmentVariableMonitor.Instance.MonitoredAppServiceEnvVarUpdatedEvent += this.UpdateEnvironmentValues;
+        }
+
+        /// <summary>
+        /// Gets or sets the <c>cloud_RoleName</c>.
+        /// </summary>
+        /// <value>The <c>cloud_RoleName</c>.</value>
+        /// <remarks>Supports environment variable expansion with <c>%VARIABLE_NAME%</c></remarks>
+        public string RoleName { get; set; }
+
+        /// <summary>
+        /// Gets or sets the <c>cloud_RoleInstance</c>.
+        /// </summary>
+        /// <value>The <c>cloud_RoleInstance</c>.</value>
+        /// <remarks>Supports environment variable expansion with <c>%VARIABLE_NAME%</c></remarks>
+        public string RoleInstance { get; set; }
+
+        /// <summary>
+        /// Initializes <see cref="ITelemetry" /> role and node context information.
+        /// </summary>
+        /// <param name="telemetry">The telemetry to initialize.</param>
+        public void Initialize(ITelemetry telemetry)
+        {
+            if (this.updateEnvVars)
+            {
+                this.roleName = this.GetRoleName();
+                this.roleInstanceName = this.GetRoleInstanceName();
+                this.updateEnvVars = false;
+            }
+
+            if (this.roleName is string && string.IsNullOrEmpty(telemetry.Context.Cloud.RoleName))
+            {
+                telemetry.Context.Cloud.RoleName = this.roleName;
+            }
+
+            if (this.roleInstanceName is string && string.IsNullOrEmpty(telemetry.Context.Cloud.RoleInstance))
+            {
+                telemetry.Context.Cloud.RoleInstance = this.roleInstanceName;
+            }
+
+            if (this.roleInstanceName is string && string.IsNullOrEmpty(telemetry.Context.GetInternalContext().NodeName))
+            {
+                telemetry.Context.GetInternalContext().NodeName = this.roleInstanceName;
+            }
+        }
+
+        /// <summary>
+        /// Remove our event handler from the environment variable monitor.
+        /// </summary>
+        public void Dispose()
+        {
+            AppServiceEnvironmentVariableMonitor.Instance.MonitoredAppServiceEnvVarUpdatedEvent -= this.UpdateEnvironmentValues;
+        }
+
+        private string GetRoleName() => ExpandEnvironmentVariables(this.RoleName);
+
+        private string GetRoleInstanceName() => ExpandEnvironmentVariables(this.RoleInstance);
+
+        private static string ExpandEnvironmentVariables(string definition)
+        {
+            if (string.IsNullOrWhiteSpace(definition))
+            {
+                return default(string);
+            }
+
+            var expandedDefinition = Regex.Replace(definition, @"(?<e>%%)|%(?<v>[^%]+)%", ExpandEnvironmentVariable, RegexOptions.ExplicitCapture);
```

Comment:
This deserves a comment (because reg-exps are hard to read). What does the replacement do?

Reply:
Also, what's wrong with `Environment.ExpandEnvironmentVariables`?

Reply:
I can add a comment, if you want to.

I have nothing against `Environment` anything. That's why I don't understand why `AppServiceEnvironmentVariableMonitor` is being used.